kube-controllers/pkg/controllers/loadbalancer/loadbalancer_controller.go

```diff
@@ -0,0 +1,915 @@
+		}
+	} else {
+		c.allocationTracker.deleteBlock(kvp.Key.String())
+		delete(c.ipamBlocks, kvp.Key.String())
+	}
+}
+
+func (c *loadBalancerController) handleIPPoolUpdate(kvp model.KVPair) {
+	if kvp.Value != nil {
+		pool := kvp.Value.(*api.IPPool)
+		if slices.Contains(pool.Spec.AllowedUses, api.IPPoolAllowedUseLoadBalancer) {
+			c.ipPools[kvp.Key.String()] = *pool
+		} else {
+			delete(c.ipPools, kvp.Key.String())
+		}
+	} else {
+		delete(c.ipPools, kvp.Key.String())
+	}
+}
+
+// syncIpam has two main uses. It functions as a garbage collection for leaked IP addresses from Service LoadBalancer
+// The other use case is to update IPs for any Service LoadBalancer that do not have IPs assigned, this could be caused by the user
+// creating Service LoadBalancer before any valid pools were created
+func (c *loadBalancerController) syncIPAM() {
+	if c.syncStatus != bapi.InSync {
+		log.WithField("status", c.syncStatus).Debug("Have not yet received InSync notification, skipping IPAM sync.")
+		return
+	}
+
+	svcObjs := make(map[serviceObject]bool)
+	services, err := c.serviceLister.Services("").List(labels.Everything())
+	if err != nil {
+		log.WithError(err).Error("Error getting service list")
+		return
+	}
+
+	for _, svc := range services {
+		if IsCalicoManagedLoadBalancer(svc, c.cfg.AssignIPs) {
+			handle, err := createHandle(svc)
+			if err != nil {
+				break
+			}
+			svcObj := serviceObject{
+				handle: handle,
+				namespacedName: types.NamespacedName{
+					Namespace: svc.Namespace,
+					Name:      svc.Name,
+				},
+			}
+			svcObjs[svcObj] = true
+		}
+	}
+
+	for _, block := range c.ipamBlocks {
+		b := block.Value.(*model.AllocationBlock)
+		for i := range b.Allocations {
+			if b.Allocations[i] != nil {
+				namespacedName := types.NamespacedName{
+					Namespace: b.Attributes[*b.Allocations[i]].AttrSecondary[ipam.AttributeNamespace],
+					Name:      b.Attributes[*b.Allocations[i]].AttrSecondary[ipam.AttributeService],
+				}
+				handle := b.Attributes[*b.Allocations[i]].AttrPrimary
+
+				svcObj := serviceObject{
+					handle:         *handle,
+					namespacedName: namespacedName,
+				}
+
+				svcObjs[svcObj] = true
+			}
+		}
+	}
+
+	for svcObj := range svcObjs {
+		err = c.syncService(svcObj)
+		if err != nil {
+			log.WithError(err).Error("Error syncing service object, will retry during next IPAM sync")
+		}
+	}
+}
+
+// syncService does the following:
+// - Releases any IP addresses in the IPAM DB associated with the Service that are not in the Service status.
+// - Allocates any addresses necessary to satisfy the Service LB request
+// - Updates the controllers internal state tracking of which IP addresses are allocated.
+// - Updates the IP addresses in the Service Status to match the IPAM DB.
+func (c *loadBalancerController) syncService(svcObj serviceObject) error {
+	svc, err := c.serviceLister.Services(svcObj.namespacedName.Namespace).Get(svcObj.namespacedName.Name)
+
+	if err != nil {
```

Comment:
I think we need to check if the error IsNotFound here - we don't want to do this for other generic errors that may occur.